lib/http/http.go

```diff
@@ -0,0 +1,391 @@
+		}
+		certpool := x509.NewCertPool()
+		pem, err := ioutil.ReadFile(opt.ClientCA)
+		if err != nil {
+			log.Fatalf("Failed to read client certificate authority: %v", err)
+			return nil, err
+		}
+		if !certpool.AppendCertsFromPEM(pem) {
+			err := errors.New("Can't parse client certificate authority")
+			log.Fatalf(err.Error())
+			return nil, err
+		}
+		tlsConfig.ClientCAs = certpool
+		tlsConfig.ClientAuth = tls.RequireAndVerifyClientCert
+	}
+
+	// Ignore passing "/" for BaseURL
+	opt.BaseURL = strings.Trim(opt.BaseURL, "/")
+	if opt.BaseURL != "" {
+		opt.BaseURL = "/" + opt.BaseURL
+	}
+
+	// Build base router
+	var router chi.Router = chi.NewRouter()
+	router.MethodNotAllowed(func(w http.ResponseWriter, _ *http.Request) {
+		http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+	})
+
+	handler := router.(http.Handler)
+	if opt.BaseURL != "" {
+		handler = http.StripPrefix(opt.BaseURL, handler)
+	}
+
+	// Serve on listeners
+	httpServer := &http.Server{
+		Handler:           handler,
+		ReadTimeout:       opt.ServerReadTimeout,
+		WriteTimeout:      opt.ServerWriteTimeout,
+		MaxHeaderBytes:    opt.MaxHeaderBytes,
+		ReadHeaderTimeout: 10 * time.Second, // time to send the headers
+		IdleTimeout:       60 * time.Second, // time to keep idle connections open
+		TLSConfig:         tlsConfig,
+	}
+
+	addrs, tlsAddrs := make([]net.Addr, len(listeners)), make([]net.Addr, len(tlsListeners))
+
+	wg := &sync.WaitGroup{}
+
+	for i, l := range listeners {
+		addrs[i] = l.Addr()
+	}
+
+	if useSSL {
+		for i, l := range tlsListeners {
+			tlsAddrs[i] = l.Addr()
+		}
+	}
+
+	return &server{addrs, tlsAddrs, listeners, tlsListeners, httpServer, router, wg, useSSL}, nil
+}
+
+func (s *server) Serve() {
+	serve := func(l net.Listener) {
+		defer s.closing.Done()
+		if err := s.httpServer.Serve(l); err != http.ErrServerClosed && err != nil {
+			log.Fatalf(err.Error())
+		}
+	}
+
+	s.closing.Add(len(s.listeners))
+	for _, l := range s.listeners {
+		go serve(l)
+	}
+
+	if s.useSSL {
+		s.closing.Add(len(s.tlsListeners))
+		for _, l := range s.tlsListeners {
+			go serve(l)
+		}
+	}
+}
+
+// Router returns the server base router
+func (s *server) Router() chi.Router {
+	return s.baseRouter
+}
+
+// Route mounts a sub-Router along a `pattern` string.
+func (s *server) Route(pattern string, fn func(r chi.Router)) chi.Router {
+	return s.baseRouter.Route(pattern, fn)
+}
+
+// Mount attaches another http.Handler along ./pattern/*
+func (s *server) Mount(pattern string, h http.Handler) {
+	s.baseRouter.Mount(pattern, h)
+}
+
+// Shutdown gracefully shuts down the server
+func (s *server) Shutdown() error {
+	if err := s.httpServer.Shutdown(context.Background()); err != nil {
+		return err
+	}
+	s.closing.Wait()
+	return nil
+}
+
+//---- Default HTTP server convenience functions ----
+
+// Router returns the server base router
+func Router() chi.Router {
+	if err := start(); err != nil {
+		panic(err.Error())
```

Comment:
I'm not mad keen on this panic. `start()` could easily return an error (eg TCP port in use).

Can we return the errors?